pkg/reduce/data_forward.go

```diff
@@ -81,164 +97,515 @@ func NewDataForward(ctx context.Context,
 		}
 	}
 
-	idleManager := wmb.NewIdleManager(len(toBuffers))
-
-	rl, err := readloop.NewReadLoop(ctx, vertexInstance.Vertex.Spec.Name, vertexInstance.Vertex.Spec.PipelineName,
-		vertexInstance.Replica, udf, pbqManager, windowingStrategy, toBuffers, whereToDecider, watermarkPublishers,
-		idleManager, options.allowedLateness)
-
-	df := &DataForward{
-		ctx:                 ctx,
-		vertexName:          vertexInstance.Vertex.Spec.Name,
-		pipelineName:        vertexInstance.Vertex.Spec.PipelineName,
-		vertexReplica:       vertexInstance.Replica,
-		fromBuffer:          fromBuffer,
-		toBuffers:           toBuffers,
-		readloop:            rl,
-		watermarkFetcher:    fw,
-		watermarkPublishers: watermarkPublishers,
-		windowingStrategy:   windowingStrategy,
-		keyed:               vertexInstance.Vertex.Spec.UDF.GroupBy.Keyed,
-		idleManager:         idleManager,
-		pbqManager:          pbqManager,
-		wmbChecker:          wmb.NewWMBChecker(2), // TODO: make configurable
-		log:                 logging.FromContext(ctx),
-		opts:                options}
-
-	return df, err
+	rl := &DataForward{
+		ctx:                   ctx,
+		vertexName:            vertexInstance.Vertex.Spec.Name,
+		pipelineName:          vertexInstance.Vertex.Spec.PipelineName,
+		vertexReplica:         vertexInstance.Replica,
+		fromBuffer:            fromBuffer,
+		toBuffers:             toBuffers,
+		watermarkFetcher:      fw,
+		watermarkPublishers:   watermarkPublishers,
+		windower:              windowingStrategy,
+		keyed:                 vertexInstance.Vertex.Spec.UDF.GroupBy.Keyed,
+		idleManager:           idleManager,
+		pbqManager:            pbqManager,
+		whereToDecider:        whereToDecider,
+		udfInvocationTracking: make(map[partition.ID]*pnf.ForwardTask),
+		of:                    of,
+		wmbChecker:            wmb.NewWMBChecker(2), // TODO: make configurable
+		log:                   logging.FromContext(ctx),
+		opts:                  options}
+
+	return rl, nil
 }
 
-// Start starts forwarding messages to readloop
-func (d *DataForward) Start() {
+// Start starts reading messages from ISG
+func (df *DataForward) Start() {
```

Comment:
```suggestion
// Start starts reading messages from ISG
func (df *DataForward) Start() {
	for {
		select {
		case <-df.ctx.Done():
			df.log.Infow("Shutting down, %s", df.ctx.Err())
			break
		default:
			// pass the child context so that the reader can be closed.
			// this way we can avoid the race condition and have all the read messages persisted
			// and acked.
			df.forwardAChunk(df.ctx)
		}
	}

	// hard shutdown after timeout
	cctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	// allow to clean itself up.
	df.ShutDown(cctx)

	return
}
```

Isn't this cleaner?
```